### PR TITLE
Exercise 4

### PR DESCRIPTION
**Pull request by powlaa**
Link to our Reviews for Team Dojo: 

https://github.com/Systems-Development-and-Frameworks/dojo/pull/6#pullrequestreview-547213463 
https://github.com/Systems-Development-and-Frameworks/dojo/pull/6#pullrequestreview-547213336

### CODE REVIEW

File: backend/.gitignore [hunk 1]
```diff
@@ -1,6 +1,7 @@
 
 .vscode/
 *.code-workspace
+.env
```
Comment (jobegrabber) [inline backend/.gitignore:4]: I'll give you a start here for
>:star: For not having security issues according to instructions

in your PR overall :smiley:

File: backend/src/datasource.js [hunk 2]
```diff
@@ -1,16 +1,27 @@
...
-    constructor(name) {
+    constructor(name, email, password) {
+        this.id = crypto.randomBytes(16).toString("hex");
         this.name = name;
+        this.email = email;
+        this.password = password;
+    }
+    static async build(name, email, password) {
+        const hashedPassword = await bcrypt.hash(password, saltRounds);
+        return new User(name, email, hashedPassword);
```
Comment (jobegrabber) [inline backend/src/datasource.js:24]: Nice!

File: backend/src/datasource.js [hunk 3]
```diff
@@ -27,40 +38,50 @@ export class InMemoryDataSource extends DataSource {
...
-    getPost(id) {
-        return Promise.resolve(this.posts.find((post) => post.id === id));
+    getPost(postId) {
+        return Promise.resolve(this.posts.find((post) => post.id === postId));
     }
 
-    createPost(data) {
-        const newPost = new Post(data);
+    createPost(title, authorId) {
+        const newPost = new Post(title, authorId);
```
Comment (jobegrabber) [inline backend/src/datasource.js:64]: +1 for explicit arguments!

File: backend/src/permissions.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+export const permissions = shield(
+    {
+        Mutation: {
+            signup: allow,
+            login: allow,
+        },
+    },
+    {
+        allowExternalErrors: true,
+        fallbackRule: isAuthenticated,
```
Comment (jobegrabber) [inline backend/src/permissions.js:16]: Could be a bit more secure with a `fallbackRule` of `deny` - but then you'll have to add a lot more rules explicitly.
OK for me

Comment (jobegrabber) [inline backend/src/permissions.js:16]: Could be a bit more secure with a `fallbackRule` of `deny` - but then you'll have to add a lot more rules explicitly.
OK for me

File: backend/src/datasource.js (hunk 3 above)
Comment (jobegrabber) [inline backend/src/datasource.js:64]: +1 for explicit arguments!

File: backend/src/datasource.js (hunk 2 above)
Comment (jobegrabber) [inline backend/src/datasource.js:24]: Nice!

File: backend/.gitignore (hunk 1 above)
Comment (jobegrabber) [inline backend/.gitignore:4]: I'll give you a start here for
>:star: For not having security issues according to instructions

in your PR overall :smiley:

File: backend/src/posts.spec.js [hunk 5]
```diff
@@ -4,12 +4,18 @@ import Server from "./server";
...
+beforeEach(async () => {
     db = new InMemoryDataSource();
-    db.users.push(new User("Jonas"));
+    await db.createUser("Jonas", "jonas@jonas.com", "Jonas1234");
+    userId = db.users[0].id;
 });
 
-const server = new Server({ dataSources: () => ({ db }) });
+const context = () => {
+    return { userId };
```
Comment (jobegrabber) [inline backend/src/posts.spec.js:15]: Great way to inject the `userId` in the tests :smiley:

File: backend/src/posts.spec.js [hunk 6]
```diff
@@ -27,20 +33,30 @@ describe("queries", () => {
...
+
+        let postQuery = () => query({ query: POSTS });
+
+        it("throws error when user is not authorised", async () => {
+            userId = null;
+            await expect(postQuery()).resolves.toMatchObject({
+                data: {
+                    posts: null,
+                },
+                errors: [expect.objectContaining({ message: "Not Authorised!" })],
```
Comment (jobegrabber) [inline backend/src/posts.spec.js:45]: I'm gonna steal this! Elegant solution to concisely match the returned object!

File: backend/src/resolvers.js [hunk 7]
```diff
@@ -1,60 +1,92 @@
...
     Mutation: {
-        write: async (parent, args, context) => {
-            const newPost = {
-                title: args.post.title,
-                authorName: args.post.author.name,
-            };
-            const author = await context.dataSources.db.getUser(newPost.authorName);
-            if (!author) {
-                throw new UserInputError("Invalid user", { invalidArgs: newPost.authorName });
+        signup: async (_, args, { dataSources }) => {
```
Comment (jobegrabber) [inline backend/src/resolvers.js:10]: I'll give you a start here 
>:star: For an implemented signup feature.

:slightly_smiling_face:

File: backend/src/resolvers.js [hunk 8]
```diff
@@ -1,60 +1,92 @@
...
             }
-            return await context.dataSources.db.createPost(newPost);
+            const users = await dataSources.db.allUsers();
+            if (users.find((user) => user.email === email)) {
+                throw new UserInputError("User with this email already exists", { invalidArgs: email });
+            }
+            const newUser = await dataSources.db.createUser(name, email, password);
+            return newUser.id;
+        },
+        login: async (_, args, { dataSources, jwt }) => {
```
Comment (jobegrabber) [inline backend/src/resolvers.js:24]: And one
> :star: For an implemented login feature using JWT.

File: backend/src/resolvers.js [hunk 9]
```diff
@@ -1,60 +1,92 @@
...
-            const author = await context.dataSources.db.getUser(newPost.authorName);
-            if (!author) {
-                throw new UserInputError("Invalid user", { invalidArgs: newPost.authorName });
+        signup: async (_, args, { dataSources }) => {
+            const name = args.name;
+            const email = args.email;
+            const password = args.password;
+            if (password.length < 9) {
+                throw new UserInputError("Password is too short", { invalidArgs: password });
             }
```
Comment (jobegrabber) [inline backend/src/resolvers.js:16]: Not exactly according to spec, but fully OK to me:
>Accept only passwords with a length of at least 8 characters.

One
>:star: For password validation.

File: backend/src/resolvers.js [hunk 10]
```diff
@@ -1,60 +1,92 @@
...
-            return await context.dataSources.db.createPost(newPost);
+            const users = await dataSources.db.allUsers();
+            if (users.find((user) => user.email === email)) {
+                throw new UserInputError("User with this email already exists", { invalidArgs: email });
```
Comment (jobegrabber) [inline backend/src/resolvers.js:19]: This here gives you a
>:star: For unique email address validation.

File: backend/src/resolvers.js [hunk 11]
```diff
@@ -1,60 +1,92 @@
...
+            const user = await dataSources.db.getUser(userId);
+            if (!user) {
+                throw new UserInputError("Invalid user", { invalidArgs: userId });
+            }
+            return await dataSources.db.createPost(title, userId);
         },
-        delete: async (parent, args, context) => {
-            const id = args.id;
-            const post = await context.dataSources.db.getPost(id);
+        upvote: async (_, args, { dataSources, userId }) => {
```
Comment (jobegrabber) [inline backend/src/resolvers.js:46]: Yet another
>:star: For assigning the authenticated user to a post in write and upvote mutations.

File: backend/src/resolvers.js [hunk 12]
```diff
@@ -63,9 +95,12 @@ const resolvers = {
...
     User: {
-        posts: async (obj, args, context) => {
-            const allPosts = await context.dataSources.db.allPosts();
-            return allPosts.filter((post) => post.authorName === obj.name);
+        posts: async (obj, args, { dataSources }) => {
+            const allPosts = await dataSources.db.allPosts();
+            return allPosts.filter((post) => post.authorId === obj.id);
+        },
+        email: (obj, args, { userId }) => {
+            return obj.id === userId ? obj.email : "";
```
Comment (jobegrabber) [inline backend/src/resolvers.js:103]: :+1:

File: backend/src/users.spec.js [hunk 13]
```diff
@@ -115,3 +134,88 @@ describe("queries", () => {
         });
     });
 });
+
+describe("mutations", () => {
+    describe("SIGNUP", () => {
```
Comment (jobegrabber) [inline backend/src/users.spec.js:139]: >:star: For a software-tested signup feature.

File: backend/src/users.spec.js [hunk 14]
```diff
@@ -115,3 +134,88 @@ describe("queries", () => {
...
+            await signupAction("Jonas", "jonas@jonas.com", "Jonas1234");
+            await expect(signupAction("Jonas", "jonas@jonas.com", "Jonas1234")).resolves.toMatchObject({
+                data: {
+                    signup: null,
+                },
+                errors: [expect.objectContaining({ message: "User with this email already exists" })],
+            });
+        });
+    });
+    describe("LOGIN", () => {
```
Comment (jobegrabber) [inline backend/src/users.spec.js:183]: >:star: For a software-tested login feature.

**Review by jobegrabber — APPROVED** [review]
Looks great!

File: backend/src/posts.spec.js (hunk 6 above)
Comment (powlaa) [reply backend/src/posts.spec.js:45]: Thought so too, Robert suggested it in the last PR 👍

File: backend/src/resolvers.js (hunk 9 above)
Comment (medizinmensch) [reply backend/src/resolvers.js:16]: Agree to both. As this only makes it more secure, it's okay I guess.

**Review by medizinmensch — APPROVED** [review]
Hey!

I can mostly confirm what @jobegrabber already wrote. 

I see that you made a review for dojo as well. And I also don't see security issues:
:star:   For requesting a review and reviewing another team's PR according to the   instructions.
:star:   For not having security issues according to instructions.

But what I see: 
image

I think one force push had an issue 😅 

But the rebase looks good I guess. So I will give you a point for that
:star: For a successful rebase   on homework/main.

Maybe you can delete or rename that branch. I guess it's as confusing for you as for me ^^

**Review by roschaefer — APPROVED** [review]
perfect

Excellent! This is perfect. I think this qualifies as a reference solution for
exercise #4.

:star: For an implemented `signup` feature.

:star: For a software-tested `signup` feature.

:star: For password validation.

:star: For unique email address validation.

:star: For not having security issues according to instructions.

:star: For an implemented `login` feature using JWT.

:star: For a software-tested `login` feature.

:star: For assigning the authenticated user to a post in `write` and `upvote` mutations.

:star: For requesting a review and reviewing another team's PR according to the instructions.

:star: For a successful rebase on `homework/main`.




There is hardly anything to improve. See very few suggestions below.

**Comment by powlaa** [comment]
@roschaefer if there are suggestions below, we can't see them yet.

**Comment by roschaefer** [comment]
> @roschaefer if there are suggestions below, we can't see them yet.

That's odd. I thought I had some. My browser doesn't have them anymore. But as far as I can remember, there weren't any important suggestions.

**Pull request merged**